### PR TITLE
Fixed for assertion failure due to not checking if we are processing Ephemeral Samples in calculate_new_heap_count

### PR DESCRIPTION
**Pull request by mrsharm**
The assertion that was failing was: ``assert (sample.elapsed_between_gcs > 0);`` and was detected by our Stress Tests by our Reliability Framework for our LOH, POH and Finalization dominant scenarios.

The fix involved checking if we are allowed to process ephemeral samples, and only if so, compute the throughput_cost_percents.

Testing: Pending and will involve rerunning the Reliability Framework Stress tests on the same scenario for ~24 hours (that's how we repro'd this issue).
Impact: This was an innocuous issue as the point where we are processing the Ephemeral samples, we do a similar check.

Call stack with assertion failure:

```
0:253> k
 # Child-SP          RetAddr               Call Site
00 00000030`f823f388 00007ffb`85d71c17     KERNELBASE!wil::details::DebugBreak+0x2
01 00000030`f823f390 00007ffb`85c737f8     coreclr!DbgAssertDialog+0x1a7 [Q:\runtime\src\coreclr\utilcode\debug.cpp @ 461] 
02 00000030`f823f480 00007ffb`85c84dbd     coreclr!SVR::gc_heap::calculate_new_heap_count+0x368 [Q:\runtime\src\coreclr\gc\gc.cpp @ 25405] 
03 00000030`f823f620 00007ffb`85c728ed     coreclr!SVR::gc_heap::gc1+0x47d [Q:\runtime\src\coreclr\gc\gc.cpp @ 22347] 
04 00000030`f823f6f0 00007ffb`85954511     coreclr!SVR::gc_heap::bgc_thread_function+0x10d [Q:\runtime\src\coreclr\gc\gc.cpp @ 39328] 
05 00000030`f823f730 00007ffc`033e257d     coreclr!<lambda_b15bb090c60fc59caad2bf8e941397d1>::operator()+0xa1 [Q:\runtime\src\coreclr\vm\gcenv.ee.cpp @ 1464] 
06 00000030`f823f760 00007ffc`04b8af28     kernel32!BaseThreadInitThunk+0x1d
07 00000030`f823f790 00000000`00000000     ntdll!RtlUserThreadStart+0x28
```

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -25402,17 +25402,21 @@ void gc_heap::calculate_new_heap_count ()
...
-        assert (sample.elapsed_between_gcs > 0);
-        throughput_cost_percents[i] = (sample.elapsed_between_gcs ? (((float)sample.msl_wait_time / n_heaps + sample.gc_pause_time) * 100.0f / (float)sample.elapsed_between_gcs) : 0.0f);
-        assert (throughput_cost_percents[i] >= 0.0);
-        if (throughput_cost_percents[i] > 100.0)
-            throughput_cost_percents[i] = 100.0;
-        dprintf (6666, ("sample %d in GC#%Id msl %I64d / %d + pause %I64d / elapsed %I64d = tcp: %.3f, surv %zd, gc speed %zd/ms", i,
-            sample.gc_index, sample.msl_wait_time, n_heaps, sample.gc_pause_time, sample.elapsed_between_gcs, throughput_cost_percents[i],
-            sample.gc_survived_size, (sample.gc_pause_time ? (sample.gc_survived_size * 1000 / sample.gc_pause_time) : 0)));
+
+    if (process_eph_samples_p)
```
Comment (mangod9) [inline src/coreclr/gc/gc.cpp:25406]: so is the `process_eph_samples_p` check the only change. weird its showing up as a larger change.

Comment (mrsharm) [reply src/coreclr/gc/gc.cpp:25406]: Correct - I think the fact we indented the for-loop, we see those as changed lines.

Comment (mrsharm) [reply src/coreclr/gc/gc.cpp:25406]: Btw, switching to hide the whitespace should highlight _just_ the added changes:

image

**Review by mangod9 — APPROVED** [review]

**Review by Maoni0 — APPROVED** [review]

**Comment by mrsharm** [comment]
Tested the scenarios by which this issue was detected from for 24 hours and found no repros => This fix worked. Merging now.

**Pull request merged**